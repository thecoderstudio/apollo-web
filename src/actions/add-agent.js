const SELECT_OPERATING_SYSTEM = 'SELECT_OPERATING_SYSTEM';
const SELECT_ARCHITECTURE = 'SELECT_ARCHITECTURE';

function selectOperatingSystem(operatingSystem) {
  return {
    type: SELECT_OPERATING_SYSTEM,
    selectedOperatingSystem: operatingSystem
  };
}

function selectArchitecture(architecture) {
  return {
    type: SELECT_ARCHITECTURE,
    selectedArchitecture: architecture
  };
}

export {
  selectOperatingSystem,
  SELECT_OPERATING_SYSTEM,
  selectArchitecture,
  SELECT_ARCHITECTURE
};
