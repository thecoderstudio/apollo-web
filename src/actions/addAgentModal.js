const SHOW_ADD_AGENT_MODAL = 'SHOW_ADD_AGENT_MODAL';
const CLOSE_ADD_AGENT_MODAL = 'CLOSE_ADD_AGENT_MODAL';

function showAddAgentModal() {
  return {
    type: SHOW_ADD_AGENT_MODAL
  };
};

function closeAddAgentModal() {
  return {
    type: SHOW_ADD_AGENT_MODAL
  };
};

export { showAddAgentModal, SHOW_ADD_AGENT_MODAL, closeAddAgentModal, CLOSE_ADD_AGENT_MODAL};
