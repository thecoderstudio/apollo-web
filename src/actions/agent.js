export const LIST_AGENTS = 'LIST_AGENTS';

export const listAgents = (agents) => ({
  type: LIST_AGENTS,
  agents
});