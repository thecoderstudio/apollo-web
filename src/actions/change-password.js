export const PROMPTED_PASSWORD_CHANGE = 'PROMPTED_PASSWORD_CHANGE';

export function promptedPasswordChange() {
  return {
    type: PROMPTED_PASSWORD_CHANGE
  };
}
