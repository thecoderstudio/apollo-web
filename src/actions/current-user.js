export const CACHE_CURRENT_USER = 'CACHE';
export const REMOVE_CURRENT_USER = 'REMOVE';
import { parseSnakeCaseObj } from '../util/parser';

function removeCurrentUser() {
  return {
    type: REMOVE_CURRENT_USER
  };
}

function cacheCurrentUser(user) {
  return {
    type: CACHE_CURRENT_USER,
    user: parseSnakeCaseObj(user)
  };
}

export { removeCurrentUser, cacheCurrentUser };
