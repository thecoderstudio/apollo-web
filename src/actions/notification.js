var id = 0

function notify(message, severity) {
  return {
    type: 'NOTIFY',
    id: id++,
    message,
    severity
  }
}

function dismiss(notificationId) {
  return {
    type: 'DISMISS',
    id: notificationId
  }
}

export { notify, dismiss };
