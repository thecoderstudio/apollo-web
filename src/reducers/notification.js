import { Map } from 'immutable';

export default function notificationReducer(state = Map({}), action) {
  switch (action.type) {
    case 'NOTIFY':
      return state.set(action.id, {
        message: action.message,
        severity: action.severity
      });
    case 'DISMISS':
      return state.delete(action.id);
    default:
      return state;
  }
}
