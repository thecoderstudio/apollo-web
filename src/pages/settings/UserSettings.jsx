import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import axios from 'axios';
import { Formik } from 'formik';
import { StatusCodes } from 'http-status-codes';
import SettingsPageWrapper from '../../components/SettingsPageWrapper';
import Button from "../../components/buttons/Button";
import { UpdateUserSchema } from '../../validation/user';
import Input from '../../components/Input';
import { fetchCurrentUser } from '../../util/user';
import { handleHTTPResponse } from '../../actions/error';
import { parseHTTPErrors } from '../../util/parser';
import { severity, notify as notifyAction } from '../../actions/notification';

const Form = styled.form`
  width: 100%;
  max-width: 300px;
`;

const StyledButton = styled(Button)`
  width: 100%;
`;

class UserSettings extends React.PureComponent {
  constructor(props) {
    super(props);
    this.updateUser = this.updateUser.bind(this);
    this.updateUserSuccessCallback = this.updateUserSuccessCallback.bind(this);
  }

  createDictionaryWithoutEmptyFields(values) {
    const data = {};
    values.username !== '' && (data.username = values.username);
    values.oldPassword !== '' && (data.old_password = values.oldPassword);
    values.password !== '' && (data.password = values.password);
    values.passwordConfirm !== '' && (data.password_confirm = values.passwordConfirm);
    return data;
  }

  updateUserSuccessCallback(resetForm) {
    const { dispatch } = this.props;

    dispatch(notifyAction("User data updated", severity.SUCCESS));
    resetForm();
    fetchCurrentUser();
  }

  updateUser(values, setErrors, resetForm) {
    axios.patch(
      `${process.env.APOLLO_HTTP_URL}user/me`,
      this.createDictionaryWithoutEmptyFields(values),
      { withCredentials: true }
    )
      .then(response => {
        this.updateUserSuccessCallback(resetForm);
      })
      .catch(error => {
        handleHTTPResponse(error.response, true, true);
        if (error.response.status === StatusCodes.BAD_REQUEST) {
          setErrors(parseHTTPErrors(error.response.data, {
            old_password: 'oldPassword',
            password_confirm: 'passwordConfirm',
            username: 'username',
            password: 'password'
          }));
        }
      });
  }

  checkForNonEmptyField(values) {
    return !Object.values(values).some(value => value !== '');
  }

  render() {
    const { className } = this.props;

    return (
      <SettingsPageWrapper {...this.props}>
        <Formik
          className={className}
          initialValues={{
            username: '',
            password: '',
            passwordConfirm: '',
            oldPassword: ''
          }}
          validationSchema={UpdateUserSchema}
          validateOnChange={false}
          onSubmit={(values, { setErrors, resetForm }) => {
            console.log(values);
            this.updateUser(values, setErrors, resetForm);
          }}
        >
          {({
            values, errors, handleChange, handleSubmit
          }) => (
            <Form onSubmit={handleSubmit}>
              <h3>Update your user settings</h3>
              <Input
                inverted
                name='username'
                placeholder='Username'
                type='username'
                value={values.username}
                error={errors.username}
                onChange={handleChange}
              />
              <Input
                inverted
                name='oldPassword'
                placeholder='Current password'
                type='password'
                value={values.oldPassword}
                error={errors.oldPassword}
                onChange={handleChange}
              />
              <Input
                inverted
                name='password'
                type='password'
                placeholder='New password'
                value={values.password}
                error={errors.password}
                onChange={handleChange}
              />
              <Input
                inverted
                name='passwordConfirm'
                type='password'
                placeholder='Confirm new password'
                value={values.passwordConfirm}
                error={errors.passwordConfirm}
                onChange={handleChange}
              />
              <StyledButton id='updateUserDetailsButton' disabled={this.checkForNonEmptyField(values)}>Update user details</StyledButton>
            </Form>
          )}
        </Formik>
      </SettingsPageWrapper>
    );
  }
}

export default connect()(UserSettings, 'User settings');

export { UserSettings as UnconnectedUserSettings };
