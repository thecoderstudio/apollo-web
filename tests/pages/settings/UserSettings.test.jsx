import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import waitForExpect from 'wait-for-expect';
import axios from 'axios';
import { StatusCodes } from 'http-status-codes';
import UserSettings from '../../../src/pages/settings/UserSettings';
import { severity, notify as notifyAction } from '../../../src/actions/notification';

const mockStore = configureStore([]);
jest.mock('axios');

function getComponent(store, props) {
  return renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <UserSettings {...props} />
      </BrowserRouter>
    </Provider>
  );
}

function submitForm(root, username, oldPassword, password, passwordConfirm) {
  const usernameInput = root.findByProps({ placeholder: 'Username' });
  const oldPasswordInput = root.findByProps({ placeholder: 'Current password' });
  const passwordInput = root.findByProps({ placeholder: 'New password' });
  const passwordConfirmInput = root.findByProps({ placeholder: 'Confirm new password' });
  const form = root.findByType('form');

  usernameInput.props.onChange({
    currentTarget: {
      name: 'username',
      value: username
    }
  });

  oldPasswordInput.props.onChange({
    currentTarget: {
      name: 'oldPassword',
      value: oldPassword
    }
  });

  passwordInput.props.onChange({
    currentTarget: {
      name: 'password',
      value: password
    }
  });

  passwordConfirmInput.props.onChange({
    currentTarget: {
      name: 'passwordConfirm',
      value: passwordConfirm
    }
  });

  form.props.onSubmit();
}

describe("User settings", () => {
  let store;
  let spy;
  const props = {
    location: {
      pathname: '/path'
    }
  };

  beforeEach(() => {
    store = mockStore({
      currentUser: {
        username: 'admin'
      }
    });
    spy = jest.spyOn(store, 'dispatch');
  });

  it("render correctly", () => {
    expect(getComponent(store, props).toJSON()).toMatchSnapshot();
  });

  it('patch successfully', async () => {
    const component = getComponent(store, props);
    const { root } = component;

    axios.get.mockResolvedValue({
      status: StatusCodes.OK,
      data: {}
    });

    axios.patch.mockResolvedValue({
      status: StatusCodes.OK
    });

    submitForm(root, 'username', 'oldpassword', 'password', 'password');

    await waitForExpect(() => {
      expect(axios.patch).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledWith(notifyAction("User data updated", severity.SUCCESS));
    });
  });

  it("validates input", async () => {
    const component = getComponent(store, props);
    const { root } = component;

    submitForm(root, 'username', '', 'password', 'password');
    submitForm(root, 'username', 'oldPassword', 'passw', 'passw');
    submitForm(root, '', 'oldpassword', 'password1', 'password2');

    await waitForExpect(() => {
      expect(axios.patch).not.toHaveBeenCalled();
    });
  });

  it('handles unsuccessful patch', async () => {
    const component = getComponent(store, props);
    const { root } = component;

    axios.patch.mockImplementationOnce(() => Promise.reject({
      response: {
        status: StatusCodes.BAD_REQUEST,
        statusText: "Bad request",
        data: {}
      }
    }));

    submitForm(root, 'username', 'oldPassword', 'password', 'password');

    axios.patch.mockImplementationOnce(() => Promise.reject({
      response: {
        status: StatusCodes.INTERNAL_SERVER_ERROR,
        statusText: "Something went wrong",
        data: {}
      }
    }));

    submitForm(root, 'username', 'oldPassword', 'password', 'password');

    await waitForExpect(() => {
      expect(axios.patch).toHaveBeenCalledTimes(2);
      expect(spy).not.toHaveBeenCalled();
    });
  });
});
