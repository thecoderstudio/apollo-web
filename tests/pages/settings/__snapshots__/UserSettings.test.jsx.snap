// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User settings render correctly 1`] = `
.c4 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
}

.c4:hover {
  opacity: 0.90;
}

.c4:focus {
  outline: none;
}

.c4:active {
  opacity: 0.80;
}

.c4:disabled:hover {
  opacity: 1;
  cursor: auto;
}

.c1 {
  height: 80px;
}

.c3 {
  min-height: 50px;
  width: 100%;
  border: none;
  border-radius: 5px;
  font-family: 'Libre Franklin',sans-serif;
  font-size: 1rem;
  padding-left: 10px;
  opacity: 0.9;
  border: 1px solid transparent;
}

.c3:focus {
  outline: none;
  border: 1px solid;
}

.c0 {
  width: 100%;
  max-width: 300px;
}

.c2 {
  margin-bottom: 15px;
}

.c5 {
  width: 100%;
}

<form
  className="c0"
  onSubmit={[Function]}
>
  <h3>
    Update your user settings
  </h3>
  <div
    className="c1 c2"
  >
    <input
      className="c3"
      name="username"
      onChange={[Function]}
      placeholder="Username"
      type="username"
      value=""
    />
  </div>
  <div
    className="c1 c2"
  >
    <input
      className="c3"
      name="oldPassword"
      onChange={[Function]}
      placeholder="Current password"
      type="password"
      value=""
    />
  </div>
  <div
    className="c1 c2"
  >
    <input
      className="c3"
      name="password"
      onChange={[Function]}
      placeholder="New password"
      type="password"
      value=""
    />
  </div>
  <div
    className="c1 c2"
  >
    <input
      className="c3"
      name="passwordConfirm"
      onChange={[Function]}
      placeholder="Confirm new password"
      type="password"
      value=""
    />
  </div>
  <button
    className="c4 c5"
    disabled={true}
    id="updateUserDetailsButton"
  >
    Update user details
  </button>
</form>
`;
